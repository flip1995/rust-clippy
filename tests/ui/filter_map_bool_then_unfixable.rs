#![allow(clippy::question_mark, unused)]
#![warn(clippy::filter_map_bool_then)]
//@no-rustfix

mod issue14368 {

    fn do_something(_: ()) -> bool {
        true
    }

    fn option_with_early_return(x: &[Option<bool>]) {
        let _ = x.iter().filter_map(|&x| x?.then(|| do_something(())));
        //~^ filter_map_bool_then
        let _ = x
            .iter()
            .filter_map(|&x| if let Some(x) = x { x } else { return None }.then(|| do_something(())));
        //~^ filter_map_bool_then
        let _ = x.iter().filter_map(|&x| {
            //~^ filter_map_bool_then
            match x {
                Some(x) => x,
                None => return None,
            }
            .then(|| do_something(()))
        });
    }

    #[derive(Copy, Clone)]
    enum Foo {
        One(bool),
        Two,
        Three(Option<i32>),
    }

    fn nested_type_with_early_return(x: &[Foo]) {
        let _ = x.iter().filter_map(|&x| {
            //~^ filter_map_bool_then
            match x {
                Foo::One(x) => x,
                Foo::Two => return None,
                Foo::Three(inner) => {
                    if inner? == 0 {
                        return Some(false);
                    } else {
                        true
                    }
                },
            }
            .then(|| do_something(()))
        });
    }
}
