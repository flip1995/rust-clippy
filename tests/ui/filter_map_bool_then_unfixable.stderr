error: usage of `bool::then` in `filter_map`
  --> tests/ui/filter_map_bool_then_unfixable.rs:12:26
   |
LL |         let _ = x.iter().filter_map(|&x| x?.then(|| do_something(())));
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider using `filter` then `map` instead
   = note: `-D clippy::filter-map-bool-then` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::filter_map_bool_then)]`

error: usage of `bool::then` in `filter_map`
  --> tests/ui/filter_map_bool_then_unfixable.rs:16:14
   |
LL |             .filter_map(|&x| if let Some(x) = x { x } else { return None }.then(|| do_something(())));
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider using `filter` then `map` instead

error: usage of `bool::then` in `filter_map`
  --> tests/ui/filter_map_bool_then_unfixable.rs:18:26
   |
LL |           let _ = x.iter().filter_map(|&x| {
   |  __________________________^
LL | |
LL | |             match x {
LL | |                 Some(x) => x,
...  |
LL | |             .then(|| do_something(()))
LL | |         });
   | |__________^
   |
   = help: consider using `filter` then `map` instead

error: usage of `bool::then` in `filter_map`
  --> tests/ui/filter_map_bool_then_unfixable.rs:36:26
   |
LL |           let _ = x.iter().filter_map(|&x| {
   |  __________________________^
LL | |
LL | |             match x {
LL | |                 Foo::One(x) => x,
...  |
LL | |             .then(|| do_something(()))
LL | |         });
   | |__________^
   |
   = help: consider using `filter` then `map` instead

error: aborting due to 4 previous errors

